### DISCUSSION TITLE
Mistweavers, Totems, and Pets

### DISCUSSION THREAD

**Discussion by yurikenus** [Hekili/hekili#1280] [Q&A] [answered]
I love Hekili and recommend it often!  I have used it extensively (with some added data) to help me learn and execute on rotations, primarily with melee healers.  I used it first with my holy paladin, and now with a mistweaver monk.  I've added I have a few questions.  I should note that I've followed a model from other class files and have added many spells and talents and used (most of) them successfully via Hekili to manage my mixed DPS&Heal rotation.

1. Is there an effective way, via this or another add-on, or external tools, to output which buffs, pets, totems, spells used, etc - via their ID?  If so, it would be a godsend.  I can mouseover a buff, or read a spellbook description, but I cannot always match it to an ID on Wowhead.  I feel like I spend more time trying to cross-reference and then test in-game than anything else.

2. I'm trying to build conditions based on the existence of Chi-ji, the Red Crane, because while he is active, I'd prioritize things differently.  I have been able to add the actual Invoke spell/talent, but I have not been able to determine if it is up as a trigger for other abilities to be prioritized.  Is Chi-ji considered a pet, or a totem, or is there a buff I can track to determine it is up?  When it is up, I will find Blackout Kick and Rising Sun Kick more valuable.  I've tried to follow what I can find in examples from other classes with pets or totems, but I've been unable to get it working.  I've also tried to see if there's a buff applied (while Chi-ji is up, the caster is immune to movement impairing effects).  No luck :(

3. Secondarily - trying to also trigger based on the "Invoke Chi-ji the Red Crane" buff which stacks and makes enveloping mist free mana/time.

So of course if there was a way to dump an ongoing log of these by ID and type, or even a snapshot of all buffs/pets/totems currently on the player, it'd be swell.  If not - any recommendations?  Maybe things that have common names (several spells, effects, talents all called "invoke_chiji_the_red_crane") need to have unique ones in Hekili's class config LUA?

**Comment by yurikenus** [comment]
I did manage to solve item 3 by a combination of trial and error, and by referencing an Weakaura's code to figure out which buff to track.

**Comment by yurikenus** [comment, marked as answer]
Well in case anyone else is working on this, I have finally found a solve, through trial and error, for item 2.  Also I'll add item 3 here for funsies.

### Getting an ID for the Chi-ji Enveloping Mists cooldown reduction
Quick note on solving item 3: it was spell ID 343820 which I added in the Auras section of the MonkMistweaver.lua so I could track against it.  I named it chiji_mists to avoid confusion with all the other Invoke Chi-ji the Red Crane references.
```
	chiji_mists = {
		id=343820,
		duration=20,
		max_stack=3,
	},
```

### Changing Priorities based on Existence of Chi-ji
I was hung up on a combination of whether things are pets or totems, and whether to identify them based on an NPC id or an icon/texture ID.  This was made worse because other classes or specs sometimes treat pets, totems, or celestials differently...  Brewmaster gets a buff while Niuzao is up which is easier to track than a totem's existence, some warlock and hunter pets are tracked as pets, and some of the files for Windwalker monk reference Xuen sometimes as a pet and sometimes as a totem.  Good times...but here's what I found.

Chi-ji the Red Crane is considered a totem - it isn't independently commandable like other pets.  My main final clue was in a Weakaura reference which just does a totem name match for Chi-ji.  I believe this is the combination of ways to use Chi-ji effectively within your action priority lists:

1. Register it as a totem with a short name and a reference to its texture ID.  I added a link like this to my MonkMistweaver.lua file:
`spec:RegisterTotem( "chiji", 877514 )`

2. Update your ability entry for Invoke Chi-ji to include a reference to the fact that you're summoning it and for how long.  I don't know all the arguments here and their meaning (wouldn't mind explanations there).  My understanding is that this basically allows Hekili's crystal ball to expect that if you do abilities in their order, abilities for 25 seconds after this one can guess Chi-ji will be around.
```
        invoke_chiji_the_red_crane = {
            id = 325197,
            cast = 0,
            cooldown = 180,
            gcd = "spell",

            spend = 0.05,
            spendType = "mana",

            toggle = "cooldowns",
            
            startsCombat = false,
            texture = 877514,

            talent = "invoke_chiji_the_red_crane",

            handler = function ()
		summonTotem( "chiji", nil, 25 )
            end,
        },
```

3. Reference Chi-ji with conditionals, if desired, when building your priority list.  For instance, when Chi-ji is up, I want to prioritize Blackout Kick higher than I would otherwise.  So I created a higher-priority entry for Blackout Kick and added this condition so Hekili knows when it should suggest Blackout Kick instead.  I added a very simple check to see if Chi-ji is active, based on what I setup above.  My conditional simply reads:
`pet.chiji.active`

**Comment by Lyvewyrez** [comment]
Hello. 

I've come here looking for a mistweaver rotation that will help make fistweaving more efficient and I'd like to know if you have anything that you can share with the rotation. I have tried to fruitless frustration to get hekili to work with mistweaver but I do not know how to use it like that. Any help would be greatly appreciated!

**Comment by yurikenus** [comment]
**Short version:** 
Healing Priority Lists aren't supported by Hekili.  My personal APLs for Mistweaver are outdated, incompatible, and unavailable.

**Longer version:** 
Healing Priority Lists aren't supported by Hekili.  Some day, damage priorities for healing specs may be supported, but APLs for healing others isn't on the roadmap.

_Why?  Well I can't speak for Hekili, but from my observations:_
The main product here is a combination of things:
- Build a flexible framework to allow an add-on to recommend stuff based on current conditions (and known future conditions)
- Build an interface to customize how this is displayed and allow some configurability
- Build a way to import the results of `simc` top priority lists in as good/baseline priority lists, making it far easier to maintain and keep current with game shifts.
- Support Classic and Retail both.
- Support all the updates and bugs along the way, including major model changes that result from major talent overhauls, tier list introductions, legendaries, and of course, new classes (which represent new energy types and mechanics like empowered spells)

Healing is hard to model well for many reasons, including (notably) the unpredictability of other players.  It may be within reach, and some have tried.  However, because targeting a group of _n_ players adds significantly more variables, recommending abilities (and targets) for those is outside the scope of Hekili, it is also probably difficult to communicate to a user, computationally expensive, and may get too close to an auto-pilot (which is bad for many reasons, too).  And of course, if there aren't good simulations to run, it means that changing priority lists (based on balance changes, power changes in gear, tier sets, trinkets, etc.) would be a very manual process, as opposed to a much slicker `simc` import that can be done for the other 35-ish specs.

_What I did:_
I have manually created some very bespoke APLs for some of my healing classes, but it required many modifications to the add-on files to work, including adding enabling the spec in Hekili and adding support for abilities, talents, buffs/auras/debuffs, and other interactions.  After that, I manually made the APLs in the Hekili UI in-game. Most of what I have done, though, you could probably do more easily with some light Weakauras to help you know which buttons are ideal under different conditions (and display them or light them up).

_My recommendation:_
Grab a Weakaura pack to get started and maybe make some modifications there if you're sometimes using the wrong spells.

**Reply by Lyvewyrez** [reply]
Yeah fair enough. I was wondering how you got it to work with Hekili even recognizing Mistweaver as a spec. 
Was happy to write my own priority-based rotation and wanted to use hekili as it is already a part of my UI and would hopefully slot in really easy. Definitely not as tech savvy as yourself - I'd have no idea how to modify hekili itself.

> Healing is hard to model well for many reasons, including (notably) the unpredictability of other players.

Intention wasn't to worry about this, just to recommend dps abilities like tiger palm, blackout kick, fae line, rising sun etc. as if it was just a dps rotation. There are some interactions and maintenance buffs like Teachings would be great to have it let me know when I need to use certain abilities. In no way was I looking at hekili to recommend healing, just the dps ability aspect.

As you say, will just have to use weak auras for now until (basic) functionality for healers is hopefully enabled in the (near?) future.

Thanks for the reply!
